metadata-ingestion/src/datahub/ingestion/source/dynamodb/dynamodb.py

```diff
@@ -0,0 +1,489 @@
+    "NS": "Number List",
+    "BS": "Binary Set",
+    "NULL": "Null",
+    # if the attribute type is NULL the attribute value will be true or false.
+    "BOOL": "Boolean",
+    "mixed": "mixed",
+}
+# map DynamoDB attribute types to DataHub classes
+_atrribute_type_to_field_type_mapping: Dict[str, Type] = {
+    "N": NumberTypeClass,
+    "B": BytesTypeClass,
+    "S": StringTypeClass,
+    "M": RecordTypeClass,
+    "L": ArrayTypeClass,
+    "SS": ArrayTypeClass,
+    "NS": ArrayTypeClass,
+    "BS": ArrayTypeClass,
+    "NULL": BooleanTypeClass,
+    "BOOL": BooleanTypeClass,
+    "mixed": UnionTypeClass,
+}
+
+
+@platform_name("DynamoDB", id="dynamodb")
+@config_class(DynamoDBConfig)
+@support_status(SupportStatus.TESTING)
+@capability(
+    SourceCapability.PLATFORM_INSTANCE,
+    "By default, platform_instance will use the AWS account id",
+)
+@capability(
+    SourceCapability.DELETION_DETECTION,
+    "Optionally enabled via `stateful_ingestion.remove_stale_metadata`",
+    supported=True,
+)
+class DynamoDBSource(StatefulIngestionSourceBase):
+    """
+    This plugin extracts the following:
+
+    AWS DynamoDB table names with their region, and infer schema of attribute names and types by scanning
+    the table
+
+    """
+
+    config: DynamoDBConfig
+    report: DynamoDBSourceReport
+    platform: str
+
+    def __init__(self, ctx: PipelineContext, config: DynamoDBConfig, platform: str):
+        super().__init__(config, ctx)
+        self.config = config
+        self.report = DynamoDBSourceReport()
+        self.platform = platform
+
+    @classmethod
+    def create(cls, config_dict: dict, ctx: PipelineContext) -> "DynamoDBSource":
+        config = DynamoDBConfig.parse_obj(config_dict)
+        return cls(ctx, config, "dynamodb")
+
+    def get_workunit_processors(self) -> List[Optional[MetadataWorkUnitProcessor]]:
+        return [
+            *super().get_workunit_processors(),
+            StaleEntityRemovalHandler.create(
+                self, self.config, self.ctx
+            ).workunit_processor,
+        ]
+
+    def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
+        # This is a offline call to get available region names from botocore library
+        session = boto3.Session()
+        dynamodb_regions = session.get_available_regions("dynamodb")
+        logger.info(f"region names {dynamodb_regions}")
+
+        # traverse databases in sorted order so output is consistent
+        for region in dynamodb_regions:
+            try:
+                # create a new dynamodb client for each region,
+                # it seems for one client we could only list the table of one specific region,
+                # the list_tables() method don't take any config that related to region
+                # TODO: list table returns maximum number 100, need to implement pagination here
+                dynamodb_client = boto3.client(
+                    "dynamodb",
+                    region_name=region,
```

Comment:
Thanks! That's a great point, Combined with the previous comment, I'm wondering if we can put this in a future improvement, not in this PR, to refactor the config to use `AwsConnectionConfig` and create a method `get_dynamodb_client` in the class to provide optional region name input. 

There's one limitation in `AwsConnectionConfig` that only accepts a single region, I think it might also make sense that we take regions as user input and loop over the regions to create clients from `get_dynamodb_client` and scan tables from there. I can put a `TODO` here for me to work on the next improvement.

Reply:
That works for me. Adding a TODO would be nice.